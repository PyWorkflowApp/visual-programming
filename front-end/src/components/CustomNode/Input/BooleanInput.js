import React, { useState, useEffect } from 'react';
import { Form } from 'react-bootstrap';

export default function BooleanInput(props) {

    const [value, setValue] = useState(props.value);
    const handleChange = (event) => {
        setValue(event.target.checked);
    };

    const {keyName, onChange} = props;
    // whenever value changes, fire callback to update config form
    useEffect(() => {
            onChange(keyName, value);
        },
        [value, keyName, onChange]);

    return  (
        <Form.Check type="checkbox" name={props.keyName}
                    disabled={props.disabled}
                    checked={value}
                    onChange={handleChange} />
    )
}
