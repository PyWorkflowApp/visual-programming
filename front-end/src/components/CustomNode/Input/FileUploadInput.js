import React, { useRef, useState, useEffect } from 'react';
import { Button } from 'react-bootstrap';
import * as API from '../../../API';

/**
 *  Component representing a file parameter.
 *  Uploads selected file to server upon selection, and passes
 *  the filename from the server response to the form callback.
 */
export default function FileUploadInput(props) {

    const input = useRef(null);
    const [fileName, setFileName] = useState(props.value || "");
    const [status, setStatus] = useState(props.value ? "ready" : "unconfigured");

    const {keyName, onChange} = props;
    // fire callback on mount to update node config state
    useEffect(() => {
            onChange(keyName, fileName);
        },
        [fileName, keyName, onChange]);

    const uploadFile = async file => {
        props.disableFunc(true);
        setStatus("loading");
        const fd = new FormData();
        fd.append("file", file);
        fd.append("nodeId", props.node.options.id);
        API.uploadDataFile(fd)
            .then(resp => {
                setFileName(resp.filename);
                setStatus("ready");
                props.disableFunc(false);
                setStatus("ready");
            }).catch(() => {
                setStatus("failed");
            });
        input.current.value = null;
    };
    const onFileSelect = e => {
        e.preventDefault();
        if (!input.current.files) {
          return;
        }
        
        uploadFile(input.current.files[0]);
    };

    if (status === "loading") return (<div>Uploading file...</div>);
    const btnText = status === "ready" ? "Choose Different File" : "Choose File";
    let content;
    if (status === "ready") {
        const rxp = new RegExp(props.node.options.id + '-');
        content = (
            <div>
                <b style={{color: 'green'}}>File loaded:</b>&nbsp;
                {fileName.replace(rxp, '')}
            </div>
        )
    } else if (status === "failed") {
        content = (<div>Upload failed. Try a new file.</div>);
    }
    return (
        <>
            <input type="file" ref={input} onChange={onFileSelect}
                   style={{display: "none"}} />
            <input type="hidden" name={props.name} value={fileName} />
            <Button size="sm" onClick={() => input.current.click()}>{btnText}</Button>
            {content}
        </>
    )
}
