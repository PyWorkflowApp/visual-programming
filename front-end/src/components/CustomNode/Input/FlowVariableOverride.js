import React from 'react';
import { Col, Form } from 'react-bootstrap';

export default function FlowVariableOverride(props) {

    const handleSelect = (event) => {
        const uuid = event.target.value;
        const flow = props.flowNodes.find(d => d.node_id === uuid);
        const obj = {
            node_id: uuid,
            is_global: flow.is_global
        };
        props.onChange(obj);
    };
    const handleCheck = (event) => { props.onFlowCheck(event.target.checked) };

    return  (
        <Col>
            <Form.Check type="checkbox" inline
                        label="Use Flow Variable"
                        checked={props.checked} onChange={handleCheck} />
            {props.checked ?
                <Form.Control as="select" name={props.keyName} onChange={handleSelect}
                              value={props.flowValue.node_id}>
                    <option/>
                    {props.flowNodes.map(fv =>
                        <option  key={fv.node_id} value={fv.node_id}>
                            {fv.options ? fv.options.var_name : fv.option_values.var_name}
                        </option>
                    )}
                </Form.Control>
                : null
            }
        </Col>
    )
}
