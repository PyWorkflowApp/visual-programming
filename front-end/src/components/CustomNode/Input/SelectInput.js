import React, { useState, useEffect } from 'react';
import { Form } from 'react-bootstrap';

export default function SelectInput(props) {

    const [value, setValue] = useState(props.value);
    const handleChange = (event) => {
        setValue(event.target.value);
    };

    const {keyName, onChange} = props;
    // whenever value changes, fire callback to update config form
    useEffect(() => {
            onChange(keyName, value);
        },
        [value, keyName, onChange]);

    return  (
        <Form.Control as="select" name={props.keyName}
                    value={value}
                    onChange={handleChange}>
            {props.options.map(opt =>
                <option key={opt} value={opt}>{opt}</option>
            )}
        </Form.Control>
    )
}
