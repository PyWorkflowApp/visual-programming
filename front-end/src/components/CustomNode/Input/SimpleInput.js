import React, { useState, useEffect } from 'react';
import { Form } from 'react-bootstrap';

export default function SimpleInput(props) {

    const [value, setValue] = useState(props.value);
    const handleChange = (event) => {
        setValue(event.target.value);
    };

    const {keyName, onChange, type} = props;
    // whenever value changes, fire callback to update config form
    useEffect(() => {
            const formValue = type === "number" ? Number(value) : value;
            onChange(keyName, formValue);
        },
        [value, keyName, onChange, type]);

    const extraProps = props.type === "textarea"
        ? {as: "textarea", rows: props.rows || 7}
        : {type: props.type};
    return (
        <Form.Control {...extraProps} name={props.keyName}
                      disabled={props.disabled}
                      defaultValue={props.value}
                      onChange={handleChange} />
    )
}
