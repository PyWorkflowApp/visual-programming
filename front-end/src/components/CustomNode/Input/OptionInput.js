import React, { useState } from 'react';
import { Col, Form } from 'react-bootstrap';
import BooleanInput from './BooleanInput'
import FileUploadInput from './FileUploadInput'
import FlowVariableOverride from './FlowVariableOverride'
import SelectInput from './SelectInput'
import SimpleInput from './SimpleInput'


/**
 *  Wrapper component to render form groups in the node config form.
 */
export default function OptionInput(props) {

    const [isFlow, setIsFlow] = useState(props.flowValue ? true : false);

    const handleFlowCheck = (bool) => {
        // if un-checking, fire callback with null so no stale value is in `option_replace`
        if (!bool) props.onChange(props.keyName, null, true);
        setIsFlow(bool);
    };

    // fire callback to update `option_replace` with flow node info
    const handleFlowVariable = (value) => {
        props.onChange(props.keyName, value, true);
    };

    let inputComp;
    if (props.type === "file") {
        inputComp = <FileUploadInput {...props} disabled={isFlow} />
    } else if (props.type === "string") {
        inputComp = <SimpleInput {...props} type="text" disabled={isFlow} />
    } else if (props.type === "text") {
        inputComp = <SimpleInput {...props} type="textarea" disabled={isFlow} />
    } else if (props.type === "int") {
        inputComp = <SimpleInput {...props} type="number" disabled={isFlow} />
    } else if (props.type === "boolean") {
        inputComp = <BooleanInput {...props} disabled={isFlow} />
    } else if (props.type === "select") {
        inputComp = <SelectInput {...props} />
    } else {
        return (<></>)
    }

    const hideFlow = props.node.options.is_global
                        || props.type === "file" || props.flowNodes.length === 0;
    return (
        <Form.Group>
            <Form.Label>{props.label}</Form.Label>
            <div className="option-docstring">{props.docstring}</div>
            <Form.Row>
                <Col xs={hideFlow ? 12 : 8}>{ inputComp }</Col>
                {hideFlow ? null :
                    <FlowVariableOverride keyName={props.keyName}
                                          flowValue={props.flowValue || {}}
                                          flowNodes={props.flowNodes || []}
                                          checked={isFlow}
                                          onFlowCheck={handleFlowCheck}
                                          onChange={handleFlowVariable} />
                }
            </Form.Row>
        </Form.Group>
    )
}
