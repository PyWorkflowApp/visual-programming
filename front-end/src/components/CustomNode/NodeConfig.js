import React from 'react';
import { Modal, Button, Form } from 'react-bootstrap';
import propTypes from 'prop-types';
import * as _ from 'lodash';
import * as API from '../../API';
import '../../styles/NodeConfig.css';
import OptionInput from './Input/OptionInput'

export default class NodeConfig extends React.Component {

    constructor(props) {
        super(props);
        this.state = {
            disabled: false,
            data: {},
            flowData: {},
            flowNodes: []
        };
        this.updateData = this.updateData.bind(this);
        this.handleDelete = this.handleDelete.bind(this);
        this.handleSubmit = this.handleSubmit.bind(this);
    }

    getFlowNodes() {
        if (!this.props.node) return;
        API.getNode(this.props.node.options.id)
            .then(node => this.setState({flowNodes: node.flow_variables}))
            .catch(err => console.log(err));
    }

    componentDidUpdate(prevProps) {
        if (!prevProps.show && this.props.show) this.getFlowNodes();
    }

    // callback to update form data in state;
    // resulting state will be sent to node config callback
    updateData(key, value, flow = false) {
        if (flow) {
            this.setState((prevState) => ({
                    ...prevState,
                    flowData: {
                        ...prevState.flowData,
                        [key]: value
                    }
                })
            );
        } else {
            this.setState((prevState) => ({
                    ...prevState,
                    data: {
                        ...prevState.data,
                        [key]: value
                    }
                })
            );
        }
    };

    // confirm, fire delete callback, close modal
    handleDelete() {
        if (window.confirm("Are you sure you want to delete this node?")) {
            this.props.onDelete();
            this.props.toggleShow();
        }
    };

    // collect config data, fire submit callback, close modal
    handleSubmit(e) {
        e.preventDefault();
        console.log(this.state.data);
        // remove items from flow vars if null
        const flowData = {...this.state.flowData};
        for (let key in flowData) {
            if (flowData[key] === null) delete flowData[key];
        }
        this.props.onSubmit(this.state.data, flowData);
        this.props.toggleShow();
    };

    render() {
        if (!this.props.node) {
          return null;
        }

        return (
            <Modal show={this.props.show} onHide={this.props.toggleShow} centered
                   dialogClassName="NodeConfig"
                   onWheel={e => e.stopPropagation()}>
                <Form onSubmit={this.handleSubmit}>
                    <Modal.Header>
                        <Modal.Title><b>{this.props.node.options.name}</b> Configuration</Modal.Title>
                    </Modal.Header>
                    <Modal.Body>
                        {_.map(this.props.node.configParams, (info, key) =>
                            <OptionInput key={key} {...info} keyName={key}
                                         onChange={this.updateData}
                                         node={this.props.node}
                                         value={this.props.node.config[key]}
                                         flowValue={this.props.node.options.option_replace ?
                                            this.props.node.options.option_replace[key] : null}
                                         flowNodes={this.state.flowNodes}
                                         disableFunc={(v) => this.setState({disabled: v})}/>
                        )}
                        <Form.Group>
                            <Form.Label>Node Description</Form.Label>
                            <Form.Control as="textarea" rows="2"
                                          name="description"
                                          onChange={e => this.updateData("description", e.target.value)}
                                          defaultValue={this.props.node.config.description}/>
                        </Form.Group>
                    </Modal.Body>
                    <Modal.Footer>
                        <Button variant="success" disabled={this.props.disabled} type="submit">Save</Button>
                        <Button variant="secondary" onClick={this.props.toggleShow}>Cancel</Button>
                        <Button variant="danger" onClick={this.handleDelete}>Delete</Button>
                    </Modal.Footer>
                </Form>
            </Modal>
        );
    }
}


NodeConfig.propTypes = {
    show: propTypes.bool,
    toggleShow: propTypes.func,
    onDelete: propTypes.func,
    onSubmit: propTypes.func
};
