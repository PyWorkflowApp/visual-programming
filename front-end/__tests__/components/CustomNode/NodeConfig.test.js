import React from 'react'
import { Col, Modal, Button, Form } from 'react-bootstrap';
import { render } from '@testing-library/react'
import { shallow, mount } from 'enzyme';
import CustomNodeModel from '../../../src/components/CustomNode/CustomNodeModel';
import NodeConfig from '../../../src/components/CustomNode/NodeConfig';
import OptionInput from '../../../src/components/CustomNode/NodeConfig';
import FileUploadInput from '../../../src/components/CustomNode/NodeConfig';
import SimpleInput from '../../../src/components/CustomNode/NodeConfig';
import BooleanInput from '../../../src/components/CustomNode/NodeConfig';
import FlowVariableOverride from '../../../src/components/CustomNode/NodeConfig';
import SelectInput from '../../../src/components/CustomNode/NodeConfig';

global.console = {log: jest.fn()}

global.fetch = jest.fn(() => Promise.resolve({
  data: [],
  json: jest.fn(() => [])
}));

global.confirm = (s) => true;

describe('Validate NodeConfig Modal', () => {
  it('Display configuration', () => {
    const node = new CustomNodeModel({id: "myId"});
    const nodeConfig = render(
      <NodeConfig node={node}
          show={true}
          toggleShow={() => {}}
          onDelete={() => {}}
          onSubmit={() => {}}  />
    );
    expect(nodeConfig).toMatchSnapshot();
  });

  it('Validates handleDelete',() => {
    const props = {
      onDelete: jest.fn(() => []),
      toggleShow: jest.fn(() => [])
    };

    const nodeConfig = new NodeConfig(props);
    nodeConfig.handleDelete();

    expect(props.onDelete.mock.calls.length).toBe(1);
    expect(props.toggleShow.mock.calls.length).toBe(1);
  });

  it('Validates handleSubmit', () => {
    const props = {
      onSubmit: jest.fn(() => []),
      toggleShow: jest.fn(() => [])
    };

    const event = {
        preventDefault: jest.fn(() => [])
    };

    const nodeConfig = new NodeConfig(props);
    nodeConfig.handleSubmit(event);

    expect(event.preventDefault.mock.calls.length).toBe(1);
    expect(props.onSubmit.mock.calls.length).toBe(1);
    expect(props.toggleShow.mock.calls.length).toBe(1);
  });

  it('Validates OptionInput with properties', () => {
    const optionInputRendered = render(<OptionInput
        keyName={"keyName"}
        node={{ options: { is_global: true }, config: {description: "Node description"}}}
        flowValue={true}
        />);

    expect(optionInputRendered).toMatchSnapshot();
  });


})
