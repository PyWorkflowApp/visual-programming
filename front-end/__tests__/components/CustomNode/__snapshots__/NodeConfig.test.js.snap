// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Validate NodeConfig Modal Display configuration 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body
    class="modal-open"
    style="padding-right: 0px;"
  >
    <div
      aria-hidden="true"
    />
    <div
      class="fade modal-backdrop show"
    />
    <div
      aria-modal="true"
      class="fade modal show"
      role="dialog"
      style="display: block;"
      tabindex="-1"
    >
      <div
        class="modal-dialog NodeConfig modal-dialog-centered"
        role="document"
      >
        <div
          class="modal-content"
        >
          <form
            class=""
          >
            <div
              class="modal-header"
            >
              <div
                class="modal-title h4"
              >
                <b />
                 Configuration
              </div>
            </div>
            <div
              class="modal-body"
            >
              <div
                class="form-group"
              >
                <label
                  class="form-label"
                >
                  Node Description
                </label>
                <textarea
                  class="form-control"
                  name="description"
                  rows="2"
                />
              </div>
            </div>
            <div
              class="modal-footer"
            >
              <button
                class="btn btn-success"
                type="submit"
              >
                Save
              </button>
              <button
                class="btn btn-secondary"
                type="button"
              >
                Cancel
              </button>
              <button
                class="btn btn-danger"
                type="button"
              >
                Delete
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  </body>,
  "container": <div
    aria-hidden="true"
  />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Validate NodeConfig Modal Validates OptionInput with properties 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body
    class=""
    style="padding-right: 0px;"
  >
    <div />
  </body>,
  "container": <div />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
